rfcs/0140-simple-package-paths.md

```diff
@@ -0,0 +1,242 @@
+- `git blame` locally and on GitHub is unaffected, since it follows file renames properly.
+- A commonly recommended way of building package directories in nixpkgs is to use `nix-build -E 'with import <nixpkgs> {}; callPackage pkgs/applications/misc/hello {}'`.
+  Since the path changes `pkg-fun.nix` is now used, this becomes like `nix-build -E 'with import <nixpkgs> {}; callPackage pkgs/unit/he/hello/pkg-fun.nix {}'`, which is harder for users.
+  However, calling a path like this is an anti-pattern anyways, because it doesn't use the correct nixpkgs version and it doesn't use the correct argument overrides.
+  The correct way of doing it was to add the package to `pkgs/top-level/all-packages.nix`, then calling `nix-build -A hello`.
+  This `nix-build -E` workaround is partially motivated by the difficulty of knowing the mapping from attributes to package paths, which is what this RFC improves upon.
+  By teaching users that `pkgs/unit/*/<name>` corresponds to `nix-build -A <name>`, the need for such `nix-build -E` workarounds should disappear.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+- The existing categorization of packages gets lost. Counter-arguments:
+  - It was never that useful to begin with
+    - The categorization was always incomplete, because packages defined in the language package sets often don't get their own categorized file path.
+    - It was an inconvenient user interface, requiring a checkout or browsing through GitHub
+    - Many packages fit multiple categories, leading to multiple locations to search through instead of one
+  - There's other better ways of discovering similar packages, e.g. [Repology](https://repology.org/)
+- This breaks `builtins.unsafeGetAttrPos "hello" pkgs`. Counter-arguments:
+  - We have to draw a line as to what constitutes the public interface of Nixpkgs. We have decided that making attribute position information part of that is not productive. For context, this information is already accepted to be unreliable at the language level, noting the `unsafe` part of the name.
+  - Support for this could be added to Nix (make `builtins.readDir` propagate file as a position)
+
+# Alternatives
+[alternatives]: #alternatives
+
+## Alternate `pkgs/unit` structure
+
+- Use a flat directory, e.g. `pkgs.hello` would be in `pkgs/unit/hello`.
+  - Good because it's simpler, both for the user and for the code
+  - Good because it speeds up Nix evaluation since there's only a single directory to call `builtins.readDir` on instead of many
+    - With an optimized `readDir` this isn't much of a problem
+  - Bad because the GitHub web interface only renders the first 1'000 entries (and we have about 10'000 that benefit from this transition, even given the restrictions)
+  - Bad because it makes `git` slower ([TODO: By how much?](https://github.com/nixpkgs-architecture/simple-package-paths/issues/18))
+  - Bad because directory listing slows down with many files
```

Comment:
Those are probably based on the same root causes and it is really hard to quantify the numbers because they depend on hardware, filesystem, caching and other running programs.

Reply:
https://github.com/NixOS/rfcs/commit/269ac4f669302c9b09075a40632b664158908a0b#diff-d795c40df63067333bb77ad1be78f1ea7e46350e124528338d970662a37f6b3bL136-R134